### PR TITLE
Add Starknet Counter example tutorial

### PR DESCRIPTION
**Pull request by devnet0x**
Add Starknet Counter example tutorial. Close issue #344 

## Summary by CodeRabbit

- **New Features**
	- Introduced the Starknet Counter application, providing a comprehensive tutorial for users.
	- Added a sidebar entry for the Starknet Counter example in the documentation.

- **Documentation**
	- Enhanced documentation with a new example tutorial for building a Starknet Counter application.

- **Bug Fixes**
	- Updated contract address in the counter service to ensure correct functionality.

### CODE REVIEW

File: examples/starknet_counter/ios/RunnerTests/RunnerTests.swift [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import UIKit
+import XCTest
+
+class RunnerTests: XCTestCase {
+
+  func testExample() {
+    // If you add code to the Runner application, consider adding tests here.
+    // See https://developer.apple.com/documentation/xctest for more information about using XCTest.
+  }
+
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/ios/RunnerTests/RunnerTests.swift:11]: this function is not used by the example

File: examples/starknet_counter/android/build.gradle [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    repositories {
+        google()
+        mavenCentral()
+    }
+
+    dependencies {
+        classpath 'com.android.tools.build:gradle:7.3.0'
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+    }
+}
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/android/build.gradle:12]: this example is intended to work in web browser  or in linux

File: examples/starknet_counter/android/build.gradle [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+rootProject.buildDir = '../build'
+subprojects {
+    project.buildDir = "${rootProject.buildDir}/${project.name}"
+}
+subprojects {
+    project.evaluationDependsOn(':app')
+}
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/android/build.gradle:27]: this is an selfcreated file by flutter, i can't change it

File: examples/starknet_counter/android/app/build.gradle [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    buildTypes {
+        release {
+            // TODO: Add your own signing config for the release build.
+            // Signing with the debug keys for now, so `flutter run --release` works.
+            signingConfig signingConfigs.debug
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/android/app/build.gradle:61]: this is an selfcreated file by flutter, i can't change it

File: examples/starknet_counter/macos/Runner.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+			name = Debug;
+		};
+		33CC111D2044C6BA0003C045 /* Release */ = {
+			isa = XCBuildConfiguration;
+			buildSettings = {
+				CODE_SIGN_STYLE = Automatic;
+				PRODUCT_NAME = "$(TARGET_NAME)";
+			};
+			name = Release;
+		};
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/macos/Runner.xcodeproj/project.pbxproj:648]: this is an selfcreated file by flutter, i can't change it

File: examples/starknet_counter/macos/Runner.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -0,0 +1,695 @@
...
+				ENABLE_NS_ASSERTIONS = NO;
+				ENABLE_STRICT_OBJC_MSGSEND = YES;
+				GCC_C_LANGUAGE_STANDARD = gnu11;
+				GCC_NO_COMMON_BLOCKS = YES;
+				GCC_WARN_64_TO_32_BIT_CONVERSION = YES;
+				GCC_WARN_ABOUT_RETURN_TYPE = YES_ERROR;
+				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
+				GCC_WARN_UNUSED_FUNCTION = YES;
+				GCC_WARN_UNUSED_VARIABLE = YES;
+				MACOSX_DEPLOYMENT_TARGET = 10.14;
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/macos/Runner.xcodeproj/project.pbxproj:460]: this is an selfcreated file by flutter, i can't change it

File: examples/starknet_counter/macos/Runner/Base.lproj/MainMenu.xib [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+                    <menu key="submenu" title="APP_NAME" systemMenu="apple" id="uQy-DD-JDr">
+                        <items>
+                            <menuItem title="About APP_NAME" id="5kV-Vb-QxS">
+                                <modifierMask key="keyEquivalentModifierMask"/>
+                                <connections>
+                                    <action selector="orderFrontStandardAboutPanel:" target="-1" id="Exp-CZ-Vem"/>
+                                </connections>
+                            </menuItem>
+                            <menuItem isSeparatorItem="YES" id="VOq-y0-SEH"/>
+                            <menuItem title="Preferences…" keyEquivalent="," id="BOF-NM-1cW"/>
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/macos/Runner/Base.lproj/MainMenu.xib:36]: this is an selfcreated file by flutter, i can't change it

File: examples/starknet_counter/macos/Runner/Base.lproj/MainMenu.xib [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+        <customObject id="Voe-Tx-rLC" customClass="AppDelegate" customModule="Runner" customModuleProvider="target">
+            <connections>
+                <outlet property="applicationMenu" destination="uQy-DD-JDr" id="XBo-yE-nKs"/>
+                <outlet property="mainFlutterWindow" destination="QvC-M9-y7g" id="gIp-Ho-8D9"/>
+            </connections>
+        </customObject>
+        <customObject id="YLy-65-1bz" customClass="NSFontManager"/>
+        <menu title="Main Menu" systemMenu="main" id="AYu-sK-qS6">
+            <items>
+                <menuItem title="APP_NAME" id="1Xt-HY-uBw">
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/macos/Runner/Base.lproj/MainMenu.xib:25]: this is an selfcreated file by flutter, i can't change it

File: examples/starknet_counter/ios/Runner.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+				ENABLE_BITCODE = NO;
+				INFOPLIST_FILE = Runner/Info.plist;
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/Frameworks",
+				);
+				PRODUCT_BUNDLE_IDENTIFIER = com.example.counter;
+				PRODUCT_NAME = "$(TARGET_NAME)";
+				SWIFT_OBJC_BRIDGING_HEADER = "Runner/Runner-Bridging-Header.h";
+				SWIFT_VERSION = 5.0;
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/ios/Runner.xcodeproj/project.pbxproj:373]: this is an selfcreated file by flutter, i can't change it

File: examples/starknet_counter/pubspec_overrides.yaml [hunk 10]
```diff
@@ -2,9 +2,13 @@ dependency_overrides:
   path: ^1.8.3
   http: ^1.0.0
   starknet:
-    path: ../../packages/starknet
+    git:
+      url: https://github.com/focustree/starknet.dart
+      path: packages/starknet
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/pubspec_overrides.yaml:7]: this is an selfcreated file by flutter, i can't change it

File: examples/starknet_counter/pubspec.yaml [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  starknet:
+  starknet_flutter:
+  flutter_lints: ^2.0.0
+  build_runner: ^2.4.6
+  starknet_provider:
+
+  # starknet_builder:
+  # wallet_kit:
+  # secure_store:
+
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/pubspec.yaml:37]: It's just a basic example for a tutorial and isn't intended for production, so  doesn't need  this contraints.

File: docs/examples/starknet-counter.mdx [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+                ElevatedButton(
+                    onPressed: _decreaseCount, child: const Text('decrement')),
+              ],
+            ),
+          ],
+        ),
+      ),
+    );
+  }
+}
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx:470]: It's just a basic example for a tutorial and isn't intended for production, so  doesn't need  this compliance.

File: examples/starknet_counter/ios/Runner.xcodeproj/project.pbxproj [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
+				CLANG_ENABLE_MODULES = YES;
+				CURRENT_PROJECT_VERSION = "$(FLUTTER_BUILD_NUMBER)";
+				ENABLE_BITCODE = NO;
+				INFOPLIST_FILE = Runner/Info.plist;
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/Frameworks",
+				);
+				PRODUCT_BUNDLE_IDENTIFIER = com.example.counter;
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/ios/Runner.xcodeproj/project.pbxproj:370]: It's just a basic example for a tutorial and isn't intended for production, so  doesn't need  this identifier.

File: examples/starknet_counter/ios/Runner.xcodeproj/project.pbxproj [hunk 14]
```diff
@@ -0,0 +1,614 @@
...
+				ENABLE_STRICT_OBJC_MSGSEND = YES;
+				GCC_C_LANGUAGE_STANDARD = gnu99;
+				GCC_NO_COMMON_BLOCKS = YES;
+				GCC_WARN_64_TO_32_BIT_CONVERSION = YES;
+				GCC_WARN_ABOUT_RETURN_TYPE = YES_ERROR;
+				GCC_WARN_UNDECLARED_SELECTOR = YES;
+				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
+				GCC_WARN_UNUSED_FUNCTION = YES;
+				GCC_WARN_UNUSED_VARIABLE = YES;
+				IPHONEOS_DEPLOYMENT_TARGET = 11.0;
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply examples/starknet_counter/ios/Runner.xcodeproj/project.pbxproj:348]: It's just a basic example for a tutorial and isn't intended for production, so  doesn't need  this contraints.

File: docs/examples/starknet-counter.mdx [hunk 15]
```diff
@@ -0,0 +1,485 @@
...
+    "class_hash": "0x61dac032f228abef9c6626f995015233097ae253a7f72d68552db02f297
+1b8f",
+    "address": "0x64b48806902a367c8598f4f95c305e8c1a1acba5f082d294a4379311311569
+1"
+  }
+}
+```
+6. Export the private key to the environment variable.
+```sh
+~/contract$ export STARKNET_PRIVATE_KEY="0x71d7bb07b9a64f6f78ac4c816aff4da9"
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx:164]: These are well know keys from developer and test environments.

File: docs/examples/starknet-counter.mdx [hunk 16]
```diff
@@ -0,0 +1,485 @@
...
+- Install starknet-devnet 0.0.7
+```sh
+asdf install starknet-devnet 0.0.7
+asdf global starknet-devnet 0.0.7
+```
+- Install starkli 0.3.5
+```sh
+asdf install starkli 0.3.5
+asdf global starkli 0.3.5
+```
```
Comment (ptisserand) [inline docs/examples/starknet-counter.mdx]: This scarb version is quite old (released in october 2023) compare to other versions

File: docs/examples/starknet-counter.mdx [hunk 17]
```diff
@@ -0,0 +1,485 @@
...
+  @override
+  State<CounterPage> createState() => _CounterPageState();
+}
+
+class _CounterPageState extends State<CounterPage> {
+  int counter = 0;
+  TextEditingController amount = TextEditingController();
+  _increaseCount() async {
+    await increaseCounter();
+    await _getCounter();
```
Comment (ptisserand) [inline docs/examples/starknet-counter.mdx:386]: Please note that In `increaseCounter` call, there is no wait for transaction acceptance, which means that `getCurrentCount` could retrieve previous value.
 Contract state will only been updated once transaction have been accepted.

File: docs/examples/starknet-counter.mdx [hunk 18]
```diff
@@ -0,0 +1,485 @@
...
+    setState(() {});
+  }
+
+  _increaseCountBy() async {
+    await increaseCounterBy(amount.text.trim());
+    await _getCounter();
```
Comment (ptisserand) [inline docs/examples/starknet-counter.mdx:392]: Same issue here

File: docs/examples/starknet-counter.mdx [hunk 19]
```diff
@@ -0,0 +1,485 @@
...
+    amount.clear();
+    setState(() {});
+  }
+
+  _decreaseCount() async {
+    await decreaseCounter();
+    await _getCounter();
```
Comment (ptisserand) [inline docs/examples/starknet-counter.mdx:399]: Same issue here

**Review by ptisserand — CHANGES_REQUESTED** [review]

File: docs/examples/starknet-counter.mdx (hunk 16 above)
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx]: updated to current version 2.8.4. Starknet devnet updated to current version too 0.2.0

File: docs/examples/starknet-counter.mdx (hunk 17 above)
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx:386]: waitForAcceptance added in counter service.

File: docs/examples/starknet-counter.mdx (hunk 18 above)
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx:392]: waitForAcceptance added in counter service.

File: docs/examples/starknet-counter.mdx (hunk 19 above)
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx:399]: waitForAcceptance added in counter service.

**Comment by devnet0x** [comment]
@ptisserand please check again, conversations solved. 😺

File: docs/examples/starknet-counter.mdx [hunk 20]
```diff
@@ -0,0 +1,492 @@
...
+final secretAccountAddress =
+    "0x64b48806902a367c8598f4f95c305e8c1a1acba5f082d294a43793113115691";
+final secretAccountPrivateKey =
+    "0x71d7bb07b9a64f6f78ac4c816aff4da9";
+final signeraccount = getAccount(
+  accountAddress: Felt.fromHexString(secretAccountAddress),
+  privateKey: Felt.fromHexString(secretAccountPrivateKey),
+  nodeUri: Uri.parse(
+      'http://localhost:5050'),
+);
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx:294]: It's only a tutorial example so doesn't need this recommendation.

File: docs/examples/starknet-counter.mdx [hunk 21]
```diff
@@ -0,0 +1,492 @@
...
+    ),
+  ]);
+
+  final txHash = response.when(
+    result: (result) => result.transaction_hash,
+    error: (err) => throw Exception("Failed to execute"),
+  );
+  print('printing decrement TX : $txHash');
+  await waitForAcceptance(transactionHash: txHash, provider: provider);
+}
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx:364]: It's only a tutorial example so doesn't need this recommendation.

File: docs/examples/starknet-counter.mdx [hunk 22]
```diff
@@ -0,0 +1,492 @@
...
+                ElevatedButton(
+                    onPressed: _decreaseCount, child: const Text('decrement')),
+              ],
+            ),
+          ],
+        ),
+      ),
+    );
+  }
+}
```
[the remark this answers is not in the document]
Comment (devnet0x) [reply docs/examples/starknet-counter.mdx:470]: It's only a tutorial example so doesn't need this recommendation.

**Review by ptisserand — APPROVED** [review]
Thanks a lot for your PR!

**Pull request merged**